addon/lib/Feature.jsm

```diff
@@ -1,140 +1,660 @@
+      await this._studyUtils.endStudy({ reason: "expired" });
+      return;
+    }
 
+    // If we're in the control group, do nothing.
+    if (this._variation.name === "control") {
+      this._log.debug("_performMeasurement - control group, bailing out");
+      return;
+    }
+
+    // Are we online? Do we have network connectivity?
+    if (this._offline()) {
+      this._log.debug("_performMeasurement - offline");
+      return;
+    }
+
+    // Check if enough time passed since the last download.
+    const lastMeasurementDate = getDateFromPref(STUDY_PREFS.LastMeasurement);
+    if (lastMeasurementDate &&
+        Math.abs(lastMeasurementDate - Date.now()) < this._delayBetweenMeasurements) {
+      this._log.debug(`_performMeasurement - skipping idle (last measurement ${lastMeasurementDate})`);
+      return;
+    }
+
+    // Bail out if a measurement is already in progress.
+    if (this._isMeasuring) {
+      this._log.warn("_performMeasurement - measurement in progress");
+      return;
+    }
+
+    // Trigger the request.
+    let endpointUrl =
+      Services.prefs.getCharPref(STUDY_PREFS.Endpoint.name, `${STUDY_PREFS.Endpoint.defaultValue}`);
+    // Append a random number to the request to bypass the cache.
+    endpointUrl = endpointUrl + "?" + (new Date()).getTime();
+
+    // Generate the request and load the script in the browser.
+    this._isMeasuring = true;
+    const url = this._triggerRequest(endpointUrl);
+    this._browser.setAttribute("src", url);
+
+    // Set the last measurement date: we don't care if it was a failure or success,
+    // as we don't want to measure more frequently than required.
+    Services.prefs.setCharPref(STUDY_PREFS.LastMeasurement.name, `${Date.now()}`);
+  }
+
+  /**
+   * Generate a custom ping and send it along with our data,
+   *
+   * @param {Object} data The full or partial data to send with the ping.
+   * @param {String} reason The reason for sending this ping, i.e. "progress"
+   *                 or "aborted".
+   */
+  async _generateAndSendPing(data, reason) {
+    // Send the measurement data.
+    let payload = {
+      reason,
+      type: data.type || "",
+      latency: data.latency,
+      fileSize: data.fileSize,
+      connType: data.connType,
+      goodput: [],
+    };
+
+    let computeGoodput = (elapsedTimeMs, downloadedInBytes) => {
+      const BYTES_PER_MEBIBYTE = 1048576;
+      const downloadedMebibytes = downloadedInBytes / BYTES_PER_MEBIBYTE;
+      const elapsedSeconds = elapsedTimeMs / 1000.0;
+      return (downloadedMebibytes * 8) / elapsedSeconds;
+    };
+
+    const checkpoints = data.transferCheckpoints || [];
+    for (let t of checkpoints) {
+      const elapsedMs = t[0];
+      payload.goodput.push([elapsedMs, computeGoodput(elapsedMs, t[1])]);
+    }
+
+    // Unfortunately, we can't simply use |this._studyUtils.telemetry| as
+    // it requires our data to be in a specific format to be serialized to
+    // the experiments parquet dataset.
```

Comment:
This is a pain point other studies have run into, so I want to double check. Are you sure you can't use the experiments schema, and are you sure you know how to process this data in Telemetry after this study runs? Using the experiments schema provides a nice "happy" path that is much easier to use than doing it yourself, but if you're comfortable going the custom route, more power to you.

Reply:
Thanks for checking on this! Yes, we know how to process the data after the custom ping hits edge servers.